### ISSUE TITLE
SOC.residual() appears to be untested and broken

### ISSUE THREAD

**Issue by rileyjmurray** [cvxpy/cvxpy#1841]
I was playing around with getting a vectorized rotated second order cone constraint in cvxpy. In the process I wanted to check feasibility of some manually assigned variable values with my implementation. When I tried to do this I got an error. 

Minimal code to reproduce:
```
import numpy as np
import cvxpy as cp
X = cp.Variable((4, 3))
X.value = np.random.randn(4, 3)
t = cp.Variable(4)
t.value = np.random.rand(4)
con = cp.SOC(t, X, axis=1)
con.violation()

  File "/Users/riley/miniforge3/envs/cvx39/lib/python3.9/site-packages/IPython/core/interactiveshell.py", line 3398, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-11-a184a5acbd60>", line 1, in <cell line: 1>
    con.violation()
  File "/Users/riley/Documents/cvxpydev/cvxpy/cvxpy/constraints/constraint.py", line 160, in violation
    residual = self.residual
  File "/Users/riley/Documents/cvxpydev/cvxpy/cvxpy/constraints/second_order.py", line 67, in residual
    zero_indices = np.where(X <= -t)[0]
ValueError: operands could not be broadcast together with shapes (4,3) (4,) 
```
Click [here](https://github.com/cvxpy/cvxpy/blob/f2b77d663922eef74cbf18b2ad0e1e2c9211f8e9/cvxpy/constraints/second_order.py#L67) to jump to the problematic line in cvxpy's codebase. I tried to fix it myself but I can't actually understand what that function is trying to accomplish.

I checked SonarCloud to see if the function was covered by our unittests, but it seems [that is not the case](https://sonarcloud.io/code?id=cvxpy_cvxpy&selected=cvxpy_cvxpy%3Acvxpy%2Fconstraints%2Fsecond_order.py&line=59).

Ping @SteveDiamond and @phschiele; we should fix this soon.

**Comment by SteveDiamond** [comment]
The general idea I believe is to project onto the SOC and then see the euclidean distance to that projection.

**Comment by rileyjmurray** [comment]
I believe the code below would fix it:
```
    @property
    def residual(self):
        """
        https://math.stackexchange.com/questions/2509986/projection-onto-the-second-order-cone
        For a scalar t and a vector x that *DO NOT* belong to the second order
        cone, the projection onto the second order cone is
            t + ||x||
            --------  *  (||x||, x).
             2||x||
        provided that || x || > abs(t).

        If ||x|| <= abs(t) but (t, x) doesn't belong to the second order
        cone then the projection is (0, 0).
        """
        t = self.args[0].value
        X = self.args[1].value
        if t is None or X is None:
            return None
        if self.axis == 0:
            X = X.T
        # Now the constraint represents np.norm(X[i,:]) <= t[i] for all i.
        norms = np.linalg.norm(X, ord=2, axis=1)
        in_cone = norms <= t

        norms_out = norms[~in_cone]
        X_out = X[~in_cone, :]
        t_out = t[~in_cone]
        tX_out = np.column_stack((t_out, X_out))

        proj_to_zero = (t_out < 0) * (norms_out <= np.abs(t_out))
        nontrivial_proj = ~proj_to_zero

        tX_proj = tX_out.copy()
        tX_proj[proj_to_zero, :] = 0
        scale = t_out[nontrivial_proj] + norms_out[nontrivial_proj]
        scale /= 2 * norms_out[nontrivial_proj]
        for i in np.where(nontrivial_proj)[0]:
            tX_proj[i, :] *= scale[i]

        distances = np.zeros(X.shape[0])
        distances[~in_cone] = np.linalg.norm(tX_out - tX_proj, axis=1)
        return distances
        ```

**Comment by SteveDiamond** [comment]
I made a PR.

**Comment by SteveDiamond** [comment]
Awkward haha, I got the projection wrong. @rileyjmurray can you update the PR with the correct projection?

**Comment by SteveDiamond** [comment]
@rileyjmurray could you fix my PR with your code?

**Comment by rileyjmurray** [comment]
Resolved by #1844.
